pkg/collector/corechecks/cluster/kubernetes_state.go

```diff
@@ -302,19 +299,11 @@ func (k *KSMCheck) hostnameAndTags(labels map[string]string, metricsToGet []ksms
 	}
 
 	// apply label joins
-	for _, mFamily := range metricsToGet {
-		config, found := k.instance.LabelJoins[mFamily.Name]
-		if !found {
-			continue
-		}
-		for _, m := range mFamily.ListMetrics {
-			if isMatching(config, labels, m.Labels) {
-				hostTag, joinedTags := k.getJoinedTags(config, m.Labels)
-				tags = append(tags, joinedTags...)
-				if hostTag != "" {
-					hostname = hostTag
-				}
-			}
+	for _, label := range labelJoiner.getLabelsToAdd(labels) {
+		tag, hostTag := k.buildTag(label.key, label.value)
+		tags = append(tags, tag)
+		if hostTag != "" {
```

Comment:
Label joiner could probably keep track of whether or not hostname override is going to happen?

Reply:
Indeed, that’s true.
The `getLabelsToAdd()` function could return both an array and a “potential” hostname.
However, I fear that it would make the code more complex.
Indeed, in order to know if a hostname override happens for a label join, the label joiner would need to leverage the `labelsMapper` setting, which it doesn’t today.
Today, `labelsMapper` is only used in `buildTag`.
And we couldn’t get rid of the “hostname override detection” in `buildTag` because `buildTag` is also used for labels not coming from the label joiner.

So, pushing the optimisation one step further by detecting hostname overrides inside the label joiner would mean that we would need two flavours of `buildTag`: one without the hostname override detection for labels coming from the label joiner and the actual one, with the hostname override detection for the labels not coming from the label joiner.

At this step, I would privilege code readability until we get new profiling data from a real deployment on a big cluster.